lib/src/git.rs

```diff
@@ -1229,6 +1229,126 @@ pub enum GitFetchError {
+    ) -> git2::FetchOptions<'_> {
+        let mut proxy_options = git2::ProxyOptions::new();
+        proxy_options.auto();
+
+        let mut fetch_options = git2::FetchOptions::new();
+        fetch_options.proxy_options(proxy_options);
+        fetch_options.remote_callbacks(callbacks.into_git());
+        if let Some(depth) = depth {
+            fetch_options.depth(depth.get().try_into().unwrap_or(i32::MAX));
+        }
+
+        fetch_options
+    }
+
+    fn fetch(
+        &mut self,
+        branch_names: &[StringPattern],
+        remote_name: &str,
+    ) -> Result<Option<String>, GitFetchError> {
+        // Perform a `git fetch` on the local git repo, updating the remote-tracking
+        // branches in the git repo.
+        let mut remote = self.git_repo.find_remote(remote_name).map_err(|err| {
+            if is_remote_not_found_err(&err) {
+                GitFetchError::NoSuchRemote(remote_name.to_string())
+            } else {
+                GitFetchError::InternalGitError(err)
+            }
+        })?;
+        // At this point, we are only updating Git's remote tracking branches, not the
+        // local branches.
+        let refspecs: Vec<_> = branch_names
+            .iter()
+            .map(|pattern| {
+                pattern
+                    .to_glob()
+                    .filter(|glob| !glob.contains(INVALID_REFSPEC_CHARS))
+                    .map(|glob| format!("+refs/heads/{glob}:refs/remotes/{remote_name}/{glob}"))
+            })
+            .collect::<Option<_>>()
+            .ok_or(GitFetchError::InvalidBranchPattern)?;
+        if refspecs.is_empty() {
+            // Don't fall back to the base refspecs.
+            return Ok(None);
+        }
+
+        tracing::debug!("remote.download");
+        remote.download(&refspecs, Some(&mut self.fetch_options))?;
+        tracing::debug!("remote.prune");
+        remote.prune(None)?;
+        tracing::debug!("remote.update_tips");
+        remote.update_tips(
+            None,
+            git2::RemoteUpdateFlags::empty(),
+            git2::AutotagOption::Unspecified,
+            None,
+        )?;
+
+        let mut default_branch = None;
+        if let Ok(default_ref_buf) = remote.default_branch() {
+            if let Some(default_ref) = default_ref_buf.as_str() {
+                // LocalBranch here is the local branch on the remote, so it's really the remote
+                // branch
+                if let Some(RefName::LocalBranch(branch_name)) = parse_git_ref(default_ref) {
+                    tracing::debug!(default_branch = branch_name);
+                    default_branch = Some(branch_name);
+                }
+            }
+        }
+        tracing::debug!("remote.disconnect");
+        remote.disconnect()?;
+        Ok(default_branch)
+    }
+
+    pub fn import_refs(
+        &mut self,
+        branch_names: &[StringPattern],
+        remote_names: &[&str],
+    ) -> Result<GitImportStats, GitImportError> {
```

Comment:
It doesn't make sense that `import_refs()` can import arbitrary refs specified by caller. My original idea was:
1. `fetch()` pushes `(branch, remote)` to `fetched: Vec<_>`
2. `import_refs()` runs `import_some_refs()` for each `fetched` elements, then clears `fetched`.

Reply:
I wasn't sure how to think about this initially, so thanks for confirming. I've updated this now. Even better, since it track {branches, remote} together, import refs maintains that mapping and doesn't do arbitrary permutations which the previous implementation was doing.